examples/skygrid.rs

```diff
@@ -0,0 +1,748 @@
+
+                if let Some(id) = &server.state {
+                    server.player_lists.get_mut(id).insert(
+                        client.uuid(),
+                        client.username(),
+                        client.textures().cloned(),
+                        client.game_mode(),
+                        0,
+                        None,
+                    );
+                }
+
+                client.send_message("Welcome to SkyGrid!".italic());
+            }
+
+            if client.is_disconnected() {
+                self.player_count.fetch_sub(1, Ordering::SeqCst);
+                if let Some(id) = &server.state {
+                    server.player_lists.get_mut(id).remove(client.uuid());
+                }
+                server.entities.remove(client.state);
+
+                return false;
+            }
+
+            if let Some(entity) = server.entities.get_mut(client.state) {
+                while handle_event_default(client, entity).is_some() {}
+            }
+
+            let dist = client.view_distance();
+            let p = client.position();
+
+            for pos in chunks_in_view_distance(ChunkPos::at(p.x, p.z), dist) {
+                if let Some(chunk) = world.chunks.get_mut(pos) {
+                    chunk.state = true;
+                } else {
+                    world.chunks.insert(pos, UnloadedChunk::default(), true);
+                }
+            }
+
+            true
+        });
+
+        // Remove chunks outside the view distance of players.
+        world.chunks.retain(|_, chunk| {
+            if chunk.state {
+                chunk.state = false;
+                true
+            } else {
+                false
+            }
+        });
+
+        // Generate chunk data for chunks created this tick.
+        world.chunks.par_iter_mut().for_each(|(pos, chunk)| {
+            if !chunk.created_this_tick() {
+                return;
+            }
+
+            for z in 0..16 {
+                for x in 0..16 {
+                    let block_x = x as i64 + pos.x as i64 * 16;
+                    let block_z = z as i64 + pos.z as i64 * 16;
+
+                    for y in (0..chunk.height()).rev() {
+                        let b = terrain_column(block_x, y as i64, block_z);
+                        chunk.set_block_state(x, y, z, b);
+                    }
+                }
+            }
+        });
+    }
+}
+
+fn terrain_column(x: i64, y: i64, z: i64) -> BlockState {
+    if has_terrain_at(x, y, z) {
+        *BLOCK_TYPES
+            .choose(&mut rand::thread_rng())
+            .unwrap_or(&BlockState::STONE)
+    } else {
+        BlockState::AIR
+    }
+}
+
+fn has_terrain_at(x: i64, y: i64, z: i64) -> bool {
+    Y_RANGE.min().unwrap_or(-64) <= y
+        && y <= Y_RANGE.max().unwrap_or(319)
+        && x % BLOCK_SPACING == 0
+        && y % BLOCK_SPACING == 0
+        && z % BLOCK_SPACING == 0
+}
+
+const BLOCK_TYPES: [BlockState; 547] = [
```

Comment:
Any chance you can get rid of this giant array? You might even be able to use `const` evaluation to compute the array at compile time, but I don't know if Rust is capable of that yet. Otherwise, it would be better to create a `Vec<BlockState>` at runtime.

Reply:
I generated that array in Python using information from `out/blocks.rs`.  I don't know how to generate the array at runtime.

Reply:
Something like this should do it:
```rust
let blocks: Vec<BlockState> = BlockKind::ALL.into_iter().map(|k| k.to_state()).filter(|b| b.is_opaque()).collect();
```

Although I think using `is_solid` would be a better heuristic to use here since this currently excludes blocks such as glass. However, Valence has not extracted this information yet, so there is no `is_solid`.

Reply:
Ok, thank you.  (It wasn't clear to me that those methods were available.)

Reply:
> ```rust
> BlockKind::ALL.into_iter().map(|k| k.to_state()).filter(|b| b.is_opaque()).collect();
> ```

This doesn't appear to be able to evaluate within a constant context, as `[BlockKind; 933]` apparently does not implement `~const IntoIterator`, so I don't know of a way to store the result in a `const` or `static` variable outside of using a crate like `once_cell`.  The alternative is to not store the value in a static way, but then I don't know where else to store the result (for use by the function `terrain_column`).

Reply:
(This issue has been resolved with the `once_cell` library.)